hbase-server/src/main/java/org/apache/hadoop/hbase/master/RegionStates.java

```diff
@@ -737,11 +741,13 @@ public void regionOffline(
   public List<HRegionInfo> serverOffline(final ZooKeeperWatcher watcher, final ServerName sn) {
     // Offline all regions on this server not already in transition.
     List<HRegionInfo> rits = new ArrayList<HRegionInfo>();
-    Set<HRegionInfo> regionsToClean = new HashSet<HRegionInfo>();
+    Set<Pair<HRegionInfo, HRegionInfo>> regionsToClean =
+      new HashSet<Pair<HRegionInfo, HRegionInfo>>();
     // Offline regions outside the loop and synchronized block to avoid
     // ConcurrentModificationException and deadlock in case of meta anassigned,
     // but RegionState a blocked.
     Set<HRegionInfo> regionsToOffline = new HashSet<HRegionInfo>();
+    Map<String, HRegionInfo> daughter2Parent = new HashMap<>();
```

Comment:
Suggest this needs more comment on why we need this accounting in daughter2Parent.